File: devices/sources/target_unstructured_data.1.cpp
/*
* @@name:	target_unstructured_data.1
* @@type:	C++
* @@operation:	compile
* @@expect:	success
* @@version:	omp_4.5
*/
class Matrix
{

  Matrix(int n) {
    len = n;
    v = new double[len];
    #pragma omp target enter data map(alloc:v[0:len])
  }

  ~Matrix() {
    // NOTE: delete map type should be used, since the corresponding 
    // host data will cease to exist after the destructor is called.

    #pragma omp target exit data map(delete:v[0:len])
    delete[] v;
  }

  private:
  double* v;
  int len;

};

File: devices/sources/target_unstructured_data.1.f90
! @@name:	target_unstructured_data.1
! @@type:	F-free
! @@operation:	compile
! @@expect:	success
! @@version:	omp_4.5
module example
  real(8), allocatable :: A(:)

  contains
    subroutine initialize(N)
      integer :: N

      allocate(A(N))
      !$omp target enter data map(alloc:A)

    end subroutine initialize

    subroutine finalize()

      !$omp target exit data map(delete:A)
      deallocate(A)

    end subroutine finalize
end module example
